source/SH3/graphics/texture.cpp

```diff
@@ -9,52 +9,382 @@
+    // Now that we're done that, we can check perform some sanity checks on our texture!
+    if(header.texSize != header.texWidth * header.texHeight * header.bpp / 8)
     {
-        Log(LogLevel::ERROR, "sh3_texture::Load( ): texWidth * texHeight * bpp != header.texSize!");
-        return;
+        Log(LogLevel::WARN, "sh3_texture::Load( ): Warning, texSize != width * height * (bpp / 8)!");
+        return; // TODO: Bind a color shader here
     }
 
-    ret = file.ReadData(&data[0], header.texSize, e);
-    if(ret != header.texSize)
-        Log(LogLevel::WARN, "sh3_texture::Load( ): Only read %d out of %d bytes!", ret, header.texSize);
+    data.resize(header.texSize); // Early data resize (if it's an 8bpp texture, it will be resized anyway)
 
-    // Now actually create a logical texture with OpenGL on the gpu
-    glGenTextures(1, &tex);
+    if(header.bpp == PixelFormat::PALETTE)
+    {
+        palette_info         pal_header;
+        std::vector<bgra>    palette;    // Palette Data (I think this is BGRA)
 
-    glBindTexture(GL_TEXTURE_2D, tex);
-    glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, header.texWidth, header.texHeight, 0, GL_BGRA, GL_UNSIGNED_BYTE, &data[0]);
-    glBindTexture(GL_TEXTURE_2D, 0); // Unbind this texture
+        // First, we need to seek to the palette and read it in.
+        file.Seek(offset + header.batchHeaderSize + header.texFileSize, std::ios_base::beg);
+        file.ReadData(&pal_header, sizeof(pal_header), e);
+
+        ASSERT(pal_header.bytes_per_pixel == sizeof(palette.front()));
+
+        // Palette information is stored in blocks (usually of size 64-bytes). We also know how large the
+        // palette is (in bytes, including padding between blocks). From this, we can deduce (with a bit of math)
+        // that the whole palette occupies ~ paletteSize/entrySize bytes of space, contains a total of
+        // entrySize/bypp colors per block, which therefore means we have a total of nBlocks * col_per_block colors,
+        // which equates to about 256-colors in total (which seems accurate for an 8-bit texture).
+
+        const std::size_t nBlocks = (pal_header.paletteSize / pal_header.entrySize) / pal_header.bytes_per_pixel;
+        const std::size_t colorsPerBlock = pal_header.entrySize / pal_header.bytes_per_pixel;
+
+        palette.resize(colorsPerBlock * nBlocks);
+
+        for(std::size_t block = 0; block < nBlocks; ++block)
+        {
+            std::size_t read = file.ReadData(static_cast<void*>(&palette[block * colorsPerBlock]), pal_header.entrySize, e);
+
+            if(read != pal_header.entrySize)
+            {
+                Log(LogLevel::WARN, "sh3_texture::Load( ): Warning: Number of bytes read in palette block != pal_header.entrySize! (Expected %d, got %d)", read, pal_header.entrySize);
+            }
+
+            // Each block is 256 bytes.
+            // Seek to next block.
+            file.Seek(256 - pal_header.entrySize, std::ios_base::cur);
```

Comment:
So you're suggesting we just leave the loop early then? Should we perhaps let the call to Load () be handled by the calling function instead of the constructor (and have Load () return an error code) so we can just bind a flat colour/checkboard shader?